src/core/pluginmanager.cpp

```diff
@@ -0,0 +1,136 @@
+/***************************************************************************
+ pluginmanager.h - PluginManager
+
+ ---------------------
+ begin                : 14.05.2024
+ copyright            : (C) 2024 by Mathieu Pellerin
+ email                : mathieu (at) opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "pluginmanager.h"
+
+#include <QDir>
+#include <QFileInfo>
+#include <QQmlApplicationEngine>
+#include <QQmlComponent>
+#include <QQmlContext>
+#include <QSettings>
+#include <qgsmessagelog.h>
+
+PluginManager::PluginManager( QQmlEngine *engine )
+  : QObject( engine )
+  , mEngine( engine )
+{
+  connect( mEngine, &QQmlEngine::warnings, this, &PluginManager::handleWarnings );
+}
+
+void PluginManager::loadPlugin( const QString &pluginPath, const QString &pluginName, bool skipPermissionCheck )
+{
+  if ( !skipPermissionCheck )
+  {
+    QSettings settings;
+    settings.beginGroup( QStringLiteral( "/qfield/plugins/%1" ).arg( pluginPath ) );
+    const QStringList keys = settings.childKeys();
+    if ( keys.contains( QStringLiteral( "permissionGranted" ) ) )
+    {
+      if ( !settings.value( QStringLiteral( "permissionGranted" ) ).toBool() )
+      {
+        return;
+      }
+    }
+    else
+    {
+      mPermissionRequestPluginPath = pluginPath;
+      emit pluginPermissionRequested( pluginName );
+      return;
+    }
+  }
+
+  if ( mLoadedPlugins.contains( pluginPath ) )
+  {
+    unloadPlugin( pluginPath );
+  }
+
+  QQmlComponent component( mEngine, pluginPath, this );
+  if ( component.status() == QQmlComponent::Status::Error )
+  {
+    for ( const QQmlError &error : component.errors() )
+    {
+      QgsMessageLog::logMessage( error.toString(), QStringLiteral( "Plugin Manager" ), Qgis::MessageLevel::Critical );
+    }
+    return;
+  }
+  QObject *object = component.create( mEngine->rootContext() );
+  mLoadedPlugins.insert( pluginPath, QPointer<QObject>( object ) );
+}
+
+void PluginManager::unloadPlugin( const QString &pluginPath )
+{
+  if ( mLoadedPlugins.contains( pluginPath ) )
+  {
+    if ( mLoadedPlugins[pluginPath] )
+    {
+      mLoadedPlugins[pluginPath]->deleteLater();
```

Comment:
Do we have a way to hook into "unload" in a plugin?

Reply:
@m-kuhn , yes, Component.onDestroyed: { ... }

Reply:
For the sake of others reading this, Component.onCompleted: { ... } is how you can hook into a plugin having been loaded.